### DISCUSSION TITLE
Remap symbols (characters)

### DISCUSSION THREAD

**Discussion by gerardbosch** [kmonad/kmonad#388] [Q&A] [answered]
First, thanks for the project, I find it great :smiley:

I was able to achieve a lot of complex and super useful stuff (home row mods, layers, shortcuts to drive workspaces/windows, launch apps,...) in a really simple manner, that's really fantastic.

My question though is about doing a more simple thing, but I'm not sure if this is out of the scope of KMonad.

Can I just remap symbols? Either in the same physical key or among keys? For example I'm interested in swapping the `;` and `:` symbols (which belong to the same key in US QWERTY). I don't want to press shift to produce `:`. And probably other similar remappings.

* In case this is not possible with KMonad as I think, what would be the recommended approach? I'm interested in Linux but also in MacOS and Windows.

I guess that at least in Linux probably `setxkbmap` (or `xmodmap`) or a script that uses that could be run when loading Linux config. In case this would be the right way to go, any advice will be very welcome :) Will `setxkbmap` work in Wayland in the future?

## The Goal
**The goal is** trying to get a universal key-map setup that can be ported among OSes (Linux/MacOS/Windows), so if I change job and I got a non-Linux computer I can keep using my custom keyboard with the less possible ceremony.

Cheers!

**Comment by m-althaf** [comment]
> I'm interested in swapping the ; and : symbols (which belong to the same key in US QWERTY). I don't want to press shift to produce :

you can combine a button like  `; (around sft ;)` or   `; (tap-hold 130 ; :)`

**Reply by MuhammedZakir** [reply]
There is shorter syntax for `(around <modifier> <button>)`. From `tutorial.kbd`:

```
  We have already seen other examples of modded buttons, \(, \), *, and +. There
  are no Keycodes for these buttons in KMonad, but they are buttons. They simply
  evaluate to `(around lsft x)`. All shifted numbers have their corresponding
  characters, the same is true for all capitals, and < > : ~ " | { } \_ + and ?.

  To wrap up 'modded-buttons', let's look back at C-A-del. We have 8 variants:
  C- : (around lctl X)
  A- : (around lalt X)
  M- : (around lmet X)
  S- : (around lsft X)

  Then RC-, RA-, RM-, and RS- behave exactly the same, except using the
  right-modifier.

  These can be combined however you please:
    C-A-M-S-x          ;; Perfectly valid
    C-%                ;; Perfectly valid: same as C-S-5
    C-RC-RA-A-M-S-RS-m ;; Sure, but why would you?
  Also, note that although we provide special syntax for certain modifiers,
  these buttons are in no way 'special' in KMonad. There is no concept of 'modifier'.
    (around a (around b c)) ;; Perfectly valid

  -------------------------------------------------------------------------- |#
```
```clojure
(defalias

  ;; Something useful
  cpy C-c
  pst C-v
  cut C-x

  ;; Something silly
  md1 (around a (around b c))    ;; abc
  md2 (around a (around lsft b)) ;; aB
  md3 C-A-M-S-l
  md4 (around % b)               ;; BEWARE: %B, not %b, do you see why?
)
```

**Reply by gerardbosch** [reply]
> you can combine a button like `; (around sft ;)` or `; (tap-hold 130 ; :)`

* I guess by `; (around sft ;)` you mean use `(around sft ;)` in `defsrc` as the definition for the `;` key, don't you?
  This will produce `:` as I want, but doing this way, I don't know how to produce `;` then :confused: 

* Regarding the trick of `(tap-hold 130 ; :)`, I don't think I can use that, as I'm using `;` key as a modded button for `ctl`:
  ```lisp
  (defalias
    ;Ctl (tap-hold-next-release 250 ; ctl)
  )
  ```

---

On the other hand I don't know if defining a dedicated `(deflayer shifted ...)` could work just to swap a few symbols (e.g. `:`/`;`, `\`/`|`,...), as I read in this issue https://github.com/kmonad/kmonad/issues/111 that it seems to be problems defining such a "shift" layer –and specially as I'm also using <kbd>s</kbd> and <kbd>l</kbd> as dual keys to emit shift when hold:

```lisp
  sSft (tap-hold-next-release 250 s sft)
  lSft (tap-hold-next-release 250 l sft :timeout-button l)
```

**Reply by m-althaf** [reply]
> * Regarding the trick of `(tap-hold 130 ; :)`, I don't think I can use that, as I'm using `;` key as a modded button for `ctl`:
>   ```lisp
>   (defalias
>     ;Ctl (tap-hold-next-release 250 ; ctl)
>   )
>   ```
I see, that seems like a whole new button which would invert shifted symbol or alphas to unshifted version of it. 
also small correction about my last comment you dont need the around() directly puting `:` in deflayer seems to give the same result

the `;` doesn't come in the home row in colemak so i use the taphold

**Reply by MuhammedZakir** [reply]
Thinking about it again, swapping keys with corresponding shifted keys don't work as expected. All the solutions posted in that issue you mentioned above and the ones I could think of have certain issues. For this to work reliably, we would need a way to "pop" pressed keys. So for now there is no "reliable" solution.

Have you tried multi-tap button? That maybe more reliable than using layers.

---

> - I guess by `; (around sft ;)` you mean use `(around sft ;)` in `defsrc` as the definition for the `;` key, don't you?
This will produce `:` as I want, but doing this way, I don't know how to produce `;` then 

You should use that in a layer, not in `defsrc`. It's better to only use "pure keys"\* in `defsrc`.

\* Keys without any modifiers.

**Reply by gerardbosch** [reply]
Sure @MuhammedZakir, I meant `deflayer` instead `defsrc` before, my fault!

Regarding multi-tap, I don't find it very ergonomic to type stuff like `:` and `;`, or `|` and `\`.

---
I'll post my findings up until now regarding this topic as a new message. I relied upon XKB to achieve what I wanted, but I'm still unsure if KMonad could handle it somehow, either now or in the future. I'm not totally happy with XB as it is platform (Linux) specific –also additional tool required besides KMonad.

**Comment by gerardbosch** [comment, marked as answer]
**UPDATE 2023-12-07:**

After a lot of time I switched from KXB to keyd because of the linked issue in XKB related with hotpluggin devices (and very happy now!). See here: https://github.com/rvaiya/keyd/issues/289#issuecomment-1846190796

---

Hi folks! Here are my findings up 'till now:

* You can totally exchange 2 keys in KMonad.
* BUT, If I'm not wrong (please, comment if I am), KMonad is not (at least right now) the right tool to swap isolated symbols (characters) either in the same key (like a character and its shifted counterpart), nor in different keys (e.g. make `shift`-`,` emmit `$`; to say something).
* Although custom remapping can be achieved by platform dependent tools.
* XKB (`xkbcomp`) will do the job on Linux, and yes! it seems it should be Wayland compatible, despite being called X.

## :penguin:  For Linux: XKB

Going through the example of swapping the `:` and `;` in US QWERTY:

File: `~/.config/autostart/xkbcomp.desktop`
- This runs `xkbcomp` and will load your keymap modifications on your desktop startup as per [Freedesktop (XDG) Autostart specification](https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html). XKB works with a display server, so can only work in graphical environments.
    ```console
    [Desktop Entry]
    Name=Load xkbcomp keyboard map modifications
    Type=Application
    # A delay seems necessary, as not doing so, xkbcomp does not take effect. I suspect because it is
    # overriden by X.org or other stuff loading the key map, so there is a race between them.
    Exec=bash -c 'sleep 3; xkbcomp -w0 -I$HOME/.config/xkb $HOME/.config/xkb/kbmap.xkb $DISPLAY'
    Terminal=false
    ```

File: `~/config/xkb/kbmap.xkb`
- This is my own setup, you can generate yours with `setxkbmap -print`.
```console
//
// Originally produced by `setxkbmap -print`, then modified to add symbol: myMods
// This keyboard map can be loaded by: `xkbcomp -I$HOME/.config/xkb kbmap $DISPLAY`
//
xkb_keymap {
    xkb_keycodes  { include "evdev+aliases(qwerty)"	};
    xkb_types     { include "complete"	};
    xkb_compat    { include "complete"	};
    xkb_symbols   { include "pc+us(altgr-intl)+es:2+gb:3+inet(evdev)+compose(menu)+terminate(ctrl_alt_bksp)+myMods"	};
    xkb_geometry  { include "pc(pc105)"	};
};
```

File: `~/.config/xkb/symbols/myMods`

```console
//
// Adding `+myMods` to `xkb_symbols` in `~/.config/xkb/kbmap` file will load this mods.
// The whole keyboard map can be inspected with `xkbcomp $DISPLAY /tmp/kbmap-dump.xkb`
//
xkb_symbols "myMods" {

    // Colon-semicolon
    key <AC10> {
        type[group1]= "FOUR_LEVEL",
        symbols[Group1]= [       colon,           semicolon,       paragraph,          degree ]
    };

    // ... other symbol swaps
};
```

### :bug: X.Org sleep/suspend issue

Loading the modified key map on desktop startup works fine, but when you suspend your computer, the changes are lost :cry: This seems to be [an old X.Org bug](https://gitlab.freedesktop.org/xorg/xserver/-/issues/266).

I could come up with [a workaround to solve that](https://gitlab.freedesktop.org/xorg/xserver/-/issues/266#note_1153904), by storing and restoring the keymap on sleep/resume.

## :question:  Open questions

Thinking in porting the above to other platforms (preferably using configuration files based tools)

1. Which would be the recommended way/tool to achieve that in MacOS? Maybe Karabiner-elements? Something easier/better?
1. Which would be the recommended way/tool to achieve that in Windows? Maybe Autohotkey? Something easier/better?
1. Is there any other tool for Linux aside XKB that works without graphical environment?
1. The final question: Should in the future KMonad be able to handle this? or this kind of redefining/swapping key maps are out of scope?

Thanks =)

**Reply by MuhammedZakir** [reply]
> 1. Which would be the recommended way/tool to achieve that in MacOS? Maybe Karabiner-elements? Something easier/better?

`hidutil`. If not possible with it, then Karabiner-Elements.

> 4. The final question: Should in the future KMonad be able to handle this? or this kind of redefining/swapping key maps are out of scope?

KMonad should support them in the future.

**Reply by gerardbosch** [reply]
Anyone on (2) and (3)? :smile:

**Reply by gerardbosch** [reply]
After thinking a little more about (3) and with the links provided by @MuhammedZakir [below](https://github.com/kmonad/kmonad/discussions/388#discussioncomment-1658407), I see more benefits of doing this with XKB than a more low level tool.

The main benefit I see is of taking benefit of XKB infrastructure to manage keyboard maps for different languages. Swapping characters via XKB can be applied on "per language" basis. I'm not very sure if other lower level tools can approach this. As an example I'm using US keyboard with a couple of remapped symbols, but once in a while I switch to stock Spanish layout: the modifications I apply for 'Group 1 en-US' in XKB does not affect 'Group 2' es.

So I think that for now XKB is quite a good option for me, despite probably not being the most simple way to go.

**Comment by molleweide** [comment]
having the ability to invert keys like OP is asking for would be amazing! I have been trying to do this in vim so that I can switch `;` and `:` but it has never succeeded LOL if this is possible it would be just so great 😃 😆

**Reply by MuhammedZakir** [reply]
I have used this in NeoVim:
```nvim
" Map ; to : and vice versa - to enter cmdline mode without shift
nnoremap ; :
nnoremap : ;
```

**Reply by molleweide** [reply]
hmmm I guess I should try it again, but this mapping hasn't worke for me before. If it works now I'm gonna feel stupid 😃

**Reply by gerardbosch** [reply]
I did this change system-wide with `xkbcomp` for now as explained above :)

**Comment by MuhammedZakir** [comment]
> 3. Is there any other tool for Linux aside XKB that works without graphical environment?

Tools that I'm aware of:
- https://github.com/rvaiya/keyd
- https://github.com/houmain/keymapper
- https://github.com/shiro/map2
- https://github.com/omakoto/keyremapper-rs
- There is also a library in Gitlab named `intercept` or `interception`. I don't have the link atm.

Edit:

- Link of `interception` tools mentioned above: https://gitlab.com/interception/linux/tools
- https://github.com/jtroo/kanata (similiar to KMonad, but implemented in Rust and doesn't support macOS (for now?)).
